File: python/tests/loop/parallel_kit.py
"""Shared pieces for the parallel tool call tests: tool bodies as coroutines, a response that
calls several tools, and a runtime whose concurrent tools are given."""

from collections import Counter
from collections.abc import Awaitable, Callable, Iterable, Sequence
from dataclasses import dataclass, field, replace
from pathlib import Path

from corpus import Clock
from kit import T0, open_store, spec, start
from pydantic import JsonValue

from threads.log import Event, JsonObject, ToolResultEvent, ToolSpec
from threads.loop.model import LookupResult, LookupUnknown
from threads.loop.runtime import Authorize, Runtime
from threads.loop.scripted import scripted_model
from threads.loop.tools import Dispatched, Invocation, Output, Termination
from threads.store import SqliteStore

type Body = Callable[[Invocation], Awaitable[Dispatched]]

USAGE: JsonValue = {"input_tokens": 5, "output_tokens": 1}
DONE: JsonValue = {
    "content": [{"type": "text", "text": "Done."}],
    "stop_reason": "end_turn",
    "usage": USAGE,
}


def calls(*names: str) -> JsonValue:
    """One response calling `names` in order, as call_1, call_2, ..."""
    parts: list[JsonValue] = [
        {"type": "tool_use", "call_id": f"call_{i + 1}", "name": n, "input": {}}
        for i, n in enumerate(names)
    ]
    return {"content": parts, "stop_reason": "tool_use", "usage": USAGE}


async def ok(call: Invocation) -> Dispatched:
    return Output(call.spec.name)


@dataclass
class Bodies:
    """A tool runner over per-tool coroutines, counting runs per tool."""

    bodies: dict[str, Body]
    runs: Counter[str] = field(default_factory=Counter[str])

    def invalid(self, spec: ToolSpec, input: JsonObject) -> str | None:
        return None

    async def dispatch(self, call: Invocation) -> Dispatched:
        self.runs[call.spec.name] += 1
        return await self.bodies[call.spec.name](call)

    async def lookup(self, call: Invocation) -> LookupResult[str]:
        return LookupUnknown("no lookup")

    async def terminate(self, call: Invocation) -> Termination:
        return "unknown"

    def provider_now(self) -> int | None:
        return None


@dataclass(frozen=True, slots=True)
class Setup:
    """A thread whose tools are `reads` (read_only) and `writes` (unguarded)."""

    reads: Sequence[str]
    writes: Sequence[str] = ()
    concurrent: Iterable[str] = ()

    def specs(self) -> list[JsonValue]:
        return [spec(n, "read_only") for n in self.reads] + [
            spec(n, "unguarded") for n in self.writes
        ]


async def begin(
    store: SqliteStore,
    setup: Setup,
    script: Sequence[JsonValue],
    tools: Bodies,
    authorize: Authorize | None = None,
) -> Runtime:
    clock = Clock(T0)
    model = scripted_model({"responses": list(script)})
    rt = await start(store, setup.specs(), model, tools, clock)
    rt = replace(rt, concurrent=frozenset(setup.concurrent))
    return rt if authorize is None else replace(rt, authorize=authorize)


async def fresh(path: Path | str = ":memory:") -> SqliteStore:
    return await open_store(path)


def results(events: Sequence[Event]) -> list[str]:
    """The call ids of the recorded results, in log order."""
    return [e.data.call_id for e in events if isinstance(e, ToolResultEvent)]

File: typescript/packages/core/test/loop/parallel-kit.ts
import { z } from "zod";
import type { KnownEvent, ToolSpec } from "../../src/log";
import type {
  LoopConfig,
  LoopEnd,
  ToolContext,
  ToolImpl,
  ToolRun,
} from "../../src/loop";
import { resume } from "../../src/loop";
import type { Writer } from "../../src/store";
import { ROOT, unwrap } from "../store/helpers";
import { events, type Harness, harness, userInput } from "./harness";

// Shared pieces for the parallel tool call tests (spec/schema/README.md, Parallel tool calls).

const usage = { input_tokens: 10, output_tokens: 2 };
export const FINAL: unknown = {
  content: [{ type: "text", text: "Done." }],
  stop_reason: "end_turn",
  usage,
};

export const read = (name: string): ToolSpec => ({
  name,
  description: `The ${name} tool.`,
  input_schema: { type: "object" },
  effect_class: "read_only",
});

export const write = (name: string): ToolSpec => ({
  ...read(name),
  effect_class: "unguarded",
});

/** One response calling `names` in order, as call_1, call_2, ... */
export function calls(...names: readonly string[]): unknown {
  return {
    content: names.map((name, i) => ({
      type: "tool_use",
      call_id: `call_${i + 1}`,
      name,
      input: {},
    })),
    stop_reason: "tool_use",
    usage,
  };
}

export const done = (output: string): ToolRun => ({
  kind: "done",
  output,
  isError: false,
});

/** A tool body; `concurrent` marks the binding like tool({concurrent: true}) does. */
export function impl(
  spec: ToolSpec,
  body: (ctx: ToolContext) => Promise<ToolRun>,
  concurrent = true,
): ToolImpl {
  return {
    spec,
    input: z.record(z.string(), z.unknown()),
    run: async (_input, ctx) => body(ctx),
    ...(concurrent ? { concurrent: true } : {}),
  };
}

export type Run = {
  readonly h: Harness;
  readonly end: LoopEnd;
  readonly log: readonly KnownEvent[];
};

/** Runs one turn on a fresh branch with these tool bodies. */
export async function run(
  impls: readonly ToolImpl[],
  responses: readonly unknown[],
  overrides: (h: Harness, writer: Writer) => Partial<LoopConfig> = () => ({}),
): Promise<Run> {
  const h = await harness(
    impls.map((i) => i.spec),
    [],
    responses,
  );
  const writer = unwrap(await h.store.acquire(ROOT, "owner", 30_000));
  const config = h.config({
    tools: new Map(impls.map((i) => [i.spec.name, i])),
    ...overrides(h, writer),
  });
  const end = await resume(writer, h.artifacts, config, {
    input: userInput("go"),
  });
  return { h, end, log: events(writer) };
}

/** Another owner takes the expired lease now. */
export async function takeOver(h: Harness): Promise<void> {
  h.clock.now += 60_000;
  unwrap(await h.store.acquire(ROOT, "usurper"));
}

export const results = (log: readonly KnownEvent[]): readonly string[] =>
  log.flatMap((e) => (e.type === "tool_result" ? [e.data.call_id] : []));

/** A promise that settles after `ms` on the real timer. */
export const after = async (ms: number): Promise<void> => {
  const { promise, resolve } = Promise.withResolvers<void>();
  setTimeout(resolve, ms);
  await promise;
};
